apps/ejabberd/src/ejabberd_c2s.erl

```diff
@@ -971,70 +970,72 @@ session_established2(El, StateData) ->
-                             resource = <<>>} ->
-                            ?DEBUG("presence_update(~p,~n\t~p,~n\t~p)",
-                                   [FromJID, PresenceEl, StateData]),
-                            presence_update(FromJID, PresenceEl,
-                                            StateData);
-                        _ ->
-                            presence_track(FromJID, ToJID, PresenceEl,
-                                           StateData)
-                    end;
-                <<"iq">> ->
-                    case jlib:iq_query_info(NewEl) of
-                        #iq{xmlns = Xmlns} = IQ
-                          when Xmlns == ?NS_PRIVACY;
-                               Xmlns == ?NS_BLOCKING ->
-                            process_privacy_iq(
-                              FromJID, ToJID, IQ, StateData);
-                        _ ->
-                            ejabberd_hooks:run(
-                              user_send_packet,
-                              Server,
-                              [FromJID, ToJID, NewEl]),
-                            check_privacy_route(FromJID, StateData, FromJID, ToJID, NewEl),
-                            StateData
-                    end;
-                <<"message">> ->
-                    ejabberd_hooks:run(user_send_packet,
-                                       Server,
-                                       [FromJID, ToJID, NewEl]),
-                    check_privacy_route(FromJID, StateData, FromJID,
-                                        ToJID, NewEl),
-                    StateData;
-                _ ->
-                    StateData
-            end
-    end,
+    NewState = process_outgoing_stanza(ToJID, Name, {Attrs, NewEl, FromJID, StateData, Server, User}),
     ejabberd_hooks:run(c2s_loop_debug, [{xmlstreamelement, El}]),
     fsm_next_state(session_established, NewState).
 
+
+process_outgoing_stanza(error, _Name, Args) ->
+    {Attrs, NewEl, _FromJID, StateData, _Server, _User} = Args,
+    case xml:get_attr_s(<<"type">>, Attrs) of
+        <<"error">> -> StateData;
+        <<"result">> -> StateData;
+        _ ->
+            Err = jlib:make_error_reply(NewEl, ?ERR_JID_MALFORMED),
+            send_element(StateData, Err),
+            StateData
+    end;
+process_outgoing_stanza(ToJID, <<"presence">>, Args) ->
+    {_Attrs, NewEl, FromJID, StateData, Server, User} = Args,
+    PresenceEl = ejabberd_hooks:run_fold(
+        c2s_update_presence,
+        Server,
+        NewEl,
+        [User, Server]),
+    ejabberd_hooks:run(
+        user_send_packet,
+        Server,
+        [FromJID, ToJID, PresenceEl]),
+    case ToJID of
+        #jid{user = User,
+            server = Server,
+            resource = <<>>} ->
+            ?DEBUG("presence_update(~p,~n\t~p,~n\t~p)",
+                [FromJID, PresenceEl, StateData]),
+            presence_update(FromJID, PresenceEl,
+                StateData);
```

Comment:
If this line has to be split, then `StateData` should be aligned to `FromJID` var.

The same goes for the debug message.